v2/internal/reconcilers/arm/arm_client_cache.go

```diff
@@ -82,37 +85,96 @@ func (c *armClientCache) lookup(key string) (*armClient, bool) {
 
 // GetClient finds and returns a client and credential to be used for a given resource
 func (c *armClientCache) GetClient(ctx context.Context, obj genruntime.ARMMetaObject) (*genericarmclient.GenericClient, string, error) {
+
+	client, err := c.checkPerResourceSecret(ctx, obj)
+	if err != nil {
+		return nil, "", err
+	} else if client != nil {
+		return client.GenericClient(), client.CredentialFrom(), nil
+	}
 	// Namespaced secret
-	secret, err := c.getSecret(ctx, obj.GetNamespace(), namespacedSecretName)
+	client, err = c.checkNamespacedSecret(ctx, obj)
 	if err != nil {
 		return nil, "", err
+	} else if client != nil {
+		return client.GenericClient(), client.CredentialFrom(), nil
+	}
+
+	// If not found, default would be global
+	return c.globalClient.GenericClient(), c.globalClient.CredentialFrom(), nil
+}
+
+func (c *armClientCache) checkNamespacedSecret(ctx context.Context, obj genruntime.ARMMetaObject) (*armClient, error) {
+	secret, err := c.getSecret(ctx, obj.GetNamespace(), namespacedSecretName)
+	if err != nil {
+		return nil, err
 	}
 
 	if secret != nil {
-		nsName := types.NamespacedName{Namespace: secret.Namespace, Name: secret.Name}
-		client, ok := c.lookup(nsName.String())
-		// if client does not already exist, or secret data is changed, we need to get new token.
-		if !ok || !matchSecretData(client.secretData, secret.Data) {
-			credential, subscriptionID, err := c.newCredentialFromSecret(secret, nsName)
-			if err != nil {
-				return nil, "", err
-			}
-
-			newClient, err := genericarmclient.NewGenericClientFromHTTPClient(c.cloudConfig, credential, c.httpClient, subscriptionID, metrics.NewARMClientMetrics())
-			if err != nil {
-				return nil, "", err
-			}
-
-			armClient := newARMClient(newClient, secret.Data, nsName)
-			c.register(armClient)
-			return armClient.GenericClient(), armClient.CredentialFrom(), nil
-		} else {
-			return client.GenericClient(), client.CredentialFrom(), nil
+		armClient, err := c.getARMClientFromSecret(secret)
+		if err != nil {
+			return nil, err
 		}
+		return armClient, nil
 	}
+	return nil, nil
+}
 
-	// If not found, default would be global
-	return c.globalClient.GenericClient(), c.globalClient.CredentialFrom(), nil
+func (c *armClientCache) checkPerResourceSecret(ctx context.Context, obj genruntime.ARMMetaObject) (*armClient, error) {
+	credentialFrom, ok := obj.GetAnnotations()[perResourceSecretAnnotation]
+	// annotation exists, use per Resource Group secret credential
+	if ok {
+		secretNamespacedName := getSecretNameFromAnnotation(credentialFrom, obj.GetNamespace())
+
+		secret, err := c.getSecret(ctx, secretNamespacedName.Namespace, secretNamespacedName.Name)
+		if err != nil {
+			return nil, err
+		}
+		// secret not found
+		if secret == nil {
+			return nil, core.NewSecretNotFoundError(secretNamespacedName, errors.Errorf("Credential Secret %q not found for '%s/%s'", secretNamespacedName.String(), obj.GetNamespace(), obj.GetName()))
```

Comment:
very minor: errors aren't supposed to start with a capital letter (Go is weird).

```suggestion
			return nil, core.NewSecretNotFoundError(secretNamespacedName, errors.Errorf("credential Secret %q not found for '%s/%s'", secretNamespacedName.String(), obj.GetNamespace(), obj.GetName()))
```

Reply:
Actually wondering if it would be cleaner to change `getSecret` to return the `NotFound` error, rather than swallowing it and returning nil.

That way you can do this:
```
		secret, err := c.getSecret(ctx, secretNamespacedName.Namespace, secretNamespacedName.Name)
		if err != nil {
			if apierrors.IsNotFound(err) {
				return nil, core.NewSecretNotFoundError(secretNamespacedName, err)
			}
			return nil, err
		}
```

The other caller (namespacedName) could do:
```
	secret, err := c.getSecret(ctx, obj.GetNamespace(), namespacedSecretName)
	if err != nil {
		if apierrors.IsNotFound(err) {
			return nil, nil // Not finding this secret is allowed, allow caller to proceed to higher scope secret
		}
		return nil, err
	}
```

Which I think reduces some extra nesting and the fact that you have to reconstruct the secret error in the per-resource secret case.